moveit_ros/moveit_servo/demos/cpp_interface/demo_pose.cpp

```diff
@@ -0,0 +1,148 @@
+#include <tf2_eigen/tf2_eigen.hpp>
+#include <moveit_servo/servo.hpp>
+#include <moveit_servo/utils.hpp>
+
+using namespace moveit_servo;
+
+namespace
+{
+const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit_servo.pose_demo");
+}
+
+int main(int argc, char* argv[])
+{
+  rclcpp::init(argc, argv);
+
+  // The servo object expects to get a ROS node.
+  auto demo_node = std::make_shared<rclcpp::Node>("moveit_servo_demo");
+
+  // Get the servo parameters.
+  std::string param_namespace = "moveit_servo";
+  auto servo_param_listener = std::make_shared<const servo::ParamListener>(demo_node, param_namespace);
+  auto servo_params = servo_param_listener->get_params();
+
+  // The publisher to send trajectory message to the robot controller.
+  rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr trajectory_outgoing_cmd_pub =
+      demo_node->create_publisher<trajectory_msgs::msg::JointTrajectory>(servo_params.command_out_topic,
+                                                                         rclcpp::SystemDefaultsQoS());
+
+  // Create the servo object
+  auto servo = Servo(demo_node, servo_param_listener);
+
+  // Wait for some time, so that we can actually see when the robot moves.
+  // This is just for convenience, should not be used for sync in real application.
+  std::this_thread::sleep_for(std::chrono::seconds(3));
+
+  // Frquency at which the commands will be send to robot controller.
+  rclcpp::WallRate rate(1.0 / servo_params.publish_period);
+
+  // Set the command type for servo.
+  servo.incomingCommandType(CommandType::POSE);
+
+  // 1.  The workflow for when command is not in planning frame.
+  Pose target_pose_ee_frame;
+  target_pose_ee_frame.frame_id = servo_params.ee_frame_name;
+  target_pose_ee_frame.pose.setIdentity();
+  // Set the target pose to +10 cm in the z direction in ee frame.
+  target_pose_ee_frame.pose.translate(Eigen::Vector3d(0.0, 0.0, 0.1));
+  // Set rotation to +45 degrees from current angle about z
+  target_pose_ee_frame.pose.rotate(Eigen::AngleAxisd(M_PI / 4, Eigen::Vector3d::UnitZ()));
+  // Servo only accepts commands in planning_frame, so use convert it before sending.
+  target_pose_ee_frame = servo.toPlanningFrame(target_pose_ee_frame);
+
+  // 2. The workflow for when command is in planning frame.
+  Pose target_pose_planning_frame;
+  target_pose_planning_frame.frame_id = servo_params.planning_frame;
+  // Set position to 10 cm in +z direction from current ee frame position in planning frame.
+  target_pose_planning_frame.pose = servo.getEndEffectorPose();
+  target_pose_planning_frame.pose.translate(Eigen::Vector3d(0.0, 0.0, -0.1));
+  // Set angle to 45 degree from current angle about z, in planning frame.
+  target_pose_planning_frame.pose.rotate(Eigen::AngleAxisd(0, Eigen::Vector3d::UnitZ()));
+
+  // Set tolerances for the target.
+  const double linear_tolerance{ 0.001 }, angular_tolerance{ 0.01 };
+
+  auto move_to_pose = [&](const auto& target_pose) {
```

Comment:
Personal preference, but this lambda is pretty long and at this point I'd prefer a regular function.

Also, should this have a timeout just in case?

Reply:
Have reduced the size of this lambda, making it into a function in the demo would require passing around 5 arguments, which felt more complex than a slightly long lambda.

Reply:
Have added a timeout as well.